### ISSUE TITLE
energy consumption and efficiency

### ISSUE THREAD

**Issue by jorgtied** [Helsinki-NLP/mammoth#106]

**Comment by amikael** [comment]
In order to analyse the objective, I first wanted to have an overview of the impact of various improvements that can possibly be made on the worst possible practice with MAMMOTH utilization.  Here is a (a little bit too long) table to present an overview. In the next comment I will focus on a selection.

Area | Do this (action) | How (one-liners / flags) | Why (effect) | Potential impact
-- | -- | -- | -- | --
Stack | Do NOT pip CUDA wheels | Don’t pip install torch or any nvidia-* deps; use LUMI PyTorch/ROCm module or supported SIF | Avoids slow/buggy fallbacks and wasted compute | High
Comms | Use OFI/CXI (aws-ofi-rccl) | Launch with srun --mpi=pmix; ensure plugin on LD_LIBRARY_PATH; set FI_PROVIDER=cxi FI_HMEM=rocr | Orders-of-magnitude faster than TCP → fewer GPU-idle seconds | High
Env | Fork-safe CXI | export CXI_FORK_SAFE=1 CXI_FORK_SAFE_HP=1 FI_CXI_DISABLE_CQ_HUGETLB=1 | Stable dataloaders / multiproc without CXI stalls | Med–High
MIOpen | Enable cache to node-local tmp | MIOPEN_DISABLE_CACHE=0; MIOPEN_USER_DB_PATH=$TMPDIR/miopen-userdb-$USER; MIOPEN_CUSTOM_CACHE_DIR=$TMPDIR/miopen-cache-$USER | Fewer repeated kernel builds → less wasted time/energy | Med
Slurm sizing | Right-size CPUs & MEM per GPU | On LUMI: ~7–8 CPUs/GPU, ~28–30 GiB/GPU (--cpus-per-task=7/8, --mem-per-gpu=30G) | Prevents over-allocation (node blocked) yet feeds dataloaders | Med
Tasks layout | 1 task per GPU | --ntasks-per-node=<gpus> --gpus-per-node=mi250:<gpus> | Keeps DDP simple, maximizes utilization | Med
I/O | Use /scratch (and /flash for metadata-heavy) | Put datasets/checkpoints on fast areas; pre-stage | Reduces dataloader stalls, improves GPU occupancy | Med
Batching | Token-based dynamic batching / bucketing | set MAMMOTH batch_type=tokens, use length buckets | Packs GPUs with similar lengths → fewer padding FLOPs | High
Seq packing | Pack multiple short samples | enable sample packing when avg length is short | Eliminates padding waste; more tokens/sec same energy | High
Precision | BF16 mixed precision | enable AMP/BF16 in config; keep loss-scaling sane | Same quality, less memory/compute → energy ↓ | High
Accumulation | Increase effective batch | tune accum_count to hit high GPU util | Better kernel efficiency without more memory | Med–High
Checkpointing | Tune frequency | save_checkpoint_steps not too small; keep_checkpoint limited | Less I/O churn, fewer syncs | Low–Med
Eval cadence | Validate less often | raise valid_steps; shrink valid_batch_size | Cuts idle/eval time during training | Low–Med
Early stop | Stop when converged | monitor BLEU/perplexity; implement patience | Avoids diminishing-returns epochs | High (when applicable)
Data quality | Filter dup/noisy pairs | dedup/clean before train | Fewer bad updates → reach target sooner | Med–High
Tokenizer | Use SPM cache & consistent vocab | keep SPM model on fast storage; reuse | Avoids re-training tokenizers; stable seq lengths | Low
Threading | Pin threads | OMP_NUM_THREADS=1–2 per worker; set num_workers | Prevents CPU thrash / context switches | Low–Med
Compilation | Try torch.compile (ROCm) | test per-model; keep a fallback | Can fuse ops → faster/greener if stable | It depends

**Comment by amikael** [comment]
Here are some areas that I recognize feasible, given what I have done, or what I can scrutinize immediately or realistically in a few months.

## Done Already

- **Do NOT pip CUDA wheels:** 
  - [MAMMOTH Tutorial](https://helsinki-nlp.github.io/mammoth/install.html): *pip does not use --site-packages* -- not yet resolved: the old instructions has to updated.  Furthermore, the installation procedure has other flaws
  - [feat/hf-integration](https://github.com/Helsinki-NLP/mammoth/tree/feat/hf_integration): asked to remove nvidia wheels from requirements.txt; now minimal: perfect! 
  - after several days study of pytorch/venv/singularity interplay - consolidated the correct installation procedure to: feat/helper: `helper/installation/build-venv.sh`

- **Use OFI/CXI (aws-ofi-rccl):** 
  - This has got no attention in [the earlier installation guide](https://helsinki-nlp.github.io/mammoth/install.html)
  - This has got no attention in [feat/hf-integration](https://github.com/Helsinki-NLP/mammoth/tree/feat/hf_integration)
  - CSC has cunfusingly earlier instructions on easybuild of aws-ofi-rccl that is no more up to date with the new ROCm
  - Using pytorch module (like on puhti) is no more what is encouraged: check the pytorch,rocm versions of the module before using, or use the latest pytorch-rocm containers.
  - Old x-transformer.py tokenizer did not allow to use newer than 3.10 python versions.  Moving away from them broke the barrier.
  - The RCCL Net Plugin `librccl-net-ofi.so` of aws-ofi-rccl was confusingly in the recent containers, but they were not loaded;   I finally managed to make RCCL to find the library.
  - The newst pytorch-rocm container does not offer enough information about aws-ofi-rccl, and what parts are internal.  When finally moving to Python 3.12, it told that the file `librccl-net-ofi.so` in the same container is not compatible but it uses now on an internal one.
  - Thus, the OFI/CXI containers are now findable, if iternal or on the LD_LIBRARY_PATH.  In testing, I encountered situations, where  LD_LIBRARY_PATH was not complete enough.   
  - The takeaway: *LD_LIBRARY_PATH and CXI environment variables are not automatically correct. The user need to check them after loading the singularity.*
  - TODO: The easies way to do this is to employ the `helper/install/build-env.sh` of the feat/helper branch as a baseline for LUMI.
  - TODO: Adapt this to Puhti.
  - TODO: Ensure that the develpement of the LD_LIBRARY_PATH setting reaches its final state asap (a couple of hours work to make sure that the setting is complete and remains as such during sbatch execution steps).


- **Fork-safe CXI**
  - After aws-ofi-rccl loading issues are sorted out, it is easy to solve this.  
  - Main LUMI [installation instructions](https://helsinki-nlp.github.io/mammoth/install.html) does not consider this; relies on  pytorch (deprecated)
  - I have reviewed some collegial practices of using singularity container.   This has not been covered there.
  - I have developed `comms-setup.sh` (insider the feat/helper branch) that contains commands to handle this with singularity.
  - On 20th September, I simplied this by adapting the old pytorch module for the newest pytorch container.
  - Now, `build-env.sh` is the key:  it sets up a software stack where a python runs automatically in the newest container (no explicit singularity call is needed) and CXI setting is implemented in the same way as in the old pytorch.
 
- **Enable cache to node-local tmp**
  - This has been studied.  Pytorch does not enable it any more, but the user is supposed to set up this.  I have addressed this need in my `task-wrapper.sh`.
 
- **Right-size CPUs & MEM per GPU: currently not solved by default**
  - The MAMMOTH installation instructions rely on the user's competence and carefulness with this.  Personally, I do not seem to learn by heart the best practice and it takes my a lot of extra time of thinking and recalling the CSC training material every time when I manually write Sbatch files.  
  - The right sizing has no not been addressed in MAMMOTH documentation - by default, you would need to take a CSC training course if you run it on CSC.  Then, learning or using the material after the training is also days of extra inefficiency.
  - If you are still confused, asking help from colleagues is useless: no one seem to be interested in these questions - without help, you need to study alone and spend another day. 
  - To address the right sizing-objective in a way that is significantly better than the current inefficient practice of MAMMOTH, I decided to write a helper utility that can suggest me a baseline that is better than probably most of use usually do without experience.   This is `mksbatch.sh`.  
  - If you are experienced in the right sizing and want to challenge this deterministic baseline, you are free to do it, but even in this case, another script, `sanity-checks.sh`, can be called to check that the automatized and your best practices are not too far from each other.
 
- **1 task per GPU**
  - This assumption is the default, also respected by my right-sizing scripts.
 
## LATER

- **Use /scratch (and /flash for metadata-heavy)**
  - I have not recognized any situation where the data of MAMMOTH would be metadata heavy: there are not too many files in normal datasets.  However, there seem to be some subtrees in /scratch/project_462000964 that have >100 000 files.  
  - TODO: I have started to look at this.  It seems that at least `MaLA-LM` file tree is huge.

- Tune frequency
- Validate less often
- Stop when converged
- Filter dup/noisy pairs
- Use SPM cache & consistent vocab

## DO NOT YET UNDERSTAND ENOUGH
 
- **BF16 mixed precision**
- Increase effective batch
- Pin threads
